modules/apps/frontend-js/frontend-js-react-web/tsconfig.json

```diff
@@ -0,0 +1,38 @@
+{
+	"@generated": "a042863abda47375e40c77a8597ebb6da0daf88d",
+	"@overrides": {
+	},
+	"@readonly": "** AUTO-GENERATED: DO NOT EDIT OUTSIDE @overrides **",
```

Comment:
So... what in this file was actually auto-generated? I guess everything other than the contents of `@overrides`? If that's the case, not sure if that's so obvious, but being json, do we have other options?

Reply:
> So... what in this file was actually auto-generated? I guess everything other than the contents of `@overrides`?

Basically. The whole file is generated by squishing three things together:

1. [Our `tsconfig-base.json`](https://github.com/liferay/liferay-frontend-projects/blob/556402b8956e8b0e7867681e865bd0e8ac362db2/projects/npm-tools/packages/npm-scripts/src/config/tsconfig-base.json) (provided by `@liferay/npm-scripts`).
2. Information about the dependency graph [derived by `@liferay/npm-scripts`](https://github.com/liferay/liferay-frontend-projects/blob/556402b8956e8b0e7867681e865bd0e8ac362db2/projects/npm-tools/packages/npm-scripts/src/typescript/getTypeScriptDependencyGraph.js) from the `dependencies` (and `devDependencies`) of our `package.json` files — this is used to populate the `paths` and `references` fields.
3. Whatever the user sticks in `@overrides`.

So, we autogenerate the entire file by doing that merge, we preserve the `@overrides` field and we do a structural comparison of the before and after (ie. ignoring the formatting). We only write the files if the structure/hash has changed, and we make sure we do it in a format that Java Source Formatter accepts, so as to avoid churn and silly `ci:test:sf` failures.

At the moment, only `remote-app-client-js` needs `@overrides`, because it is not actually a `liferay-portal` project per se, and exists just to build the standalone SDK via a different process.

> do we have other options?

One nice thing about TS is that it allows comments in JSON files (ie. they are not really JSON). Prettier does too. So that means we could do something like:

```
// @generated DO NOT EDIT — HASH HERE WHATEVER etc
{
    "the real config": "here"
}
```

However, we decided that we wanted Java Source Formatter to be responsible for all `.json` files, so that rules out the possibility of adding comments.